ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -260,10 +260,19 @@ densityDisconnect (GenesisWindow sgen) (SecurityParam k) states candidateSuffixe
       pure (peer, DensityBounds {fragment, offersMoreThanK, lowerBound, upperBound, hasBlockAfter, latestSlot, idling})
 
     losingPeers = nubOrd $ do
-      (peer0 , DensityBounds {fragment = frag0, upperBound = ub0}) <-
+      (peer0 , DensityBounds { fragment = frag0
+                             , upperBound = ub0
+                             , hasBlockAfter = hasBlockAfter0
+                             , idling = idling0
+                             }) <-
         Map.toList densityBounds
       (_peer1, DensityBounds {fragment = frag1, offersMoreThanK, lowerBound = lb1 }) <-
         Map.toList densityBounds
+      -- Don't disconnect peer0 if it sent no headers after the intersection yet
+      -- and it is not idling.
+      --
+      -- See Note [Chain disagreement]
+      guard $ idling0 || not (AF.null frag0) || hasBlockAfter0
       -- ensure that the two peer fragments don't share any
       -- headers after the LoE
       guard $ AF.lastPoint frag0 /= AF.lastPoint frag1
```

Comment:
Ah right, when we disconnect from someone due to `lb1 == ub0`, we need to be sure that they actually committed to fragments such that their intersection is the LoE tip, which is *not* guaranteed by the existing `AF.lastPoint frag0 /= AF.lastPoint frag1`.

Maybe we could extract this logic (these two `guard`s) to a separate function `definitelyIntersectAtLoE` and test it, as the logic isn't immediately obvious?

Reply:
Hmm, what happens when both the dynamo and the objector serve a header after the forecast horizon? Ideally, the GDD should disconnect from both of them, but I don't think it does currently, as `guard $ AF.lastPoint frag0 /= AF.lastPoint frag1` will prevent that. This seems to be an exception to the rule "the fragments have to disagree after the intersection".

Maybe that case is handled in an ad-hoc fashion in the CSJ logic though; but it seems good to be robust here.

Reply:
> Hmm, what happens when both the dynamo and the objector serve a header after the forecast horizon?

I think this is a good point that deserves another PR.

Reply:
> Maybe we could extract this logic (these two guards) to a separate function definitelyIntersectAtLoE and test it, as the logic isn't immediately obvious?

I'd be happy to review some proposal to do that.